src/build.c

```diff
@@ -662,61 +655,229 @@ static void clear_all_errors(void)
 }
 
 
-/* Replaces occurrences of %e and %p with the appropriate filenames and
- * %l with current line number. %d and %p replacements should be in UTF8 */
-static gchar *build_replace_placeholder(const GeanyDocument *doc, const gchar *src)
+typedef struct
+{
+	const gchar placeholder;
+	const gchar *replacement;
+} Placeholder;
+
+
+static const gchar *placeholder_replacement(Placeholder *placeholders, size_t n_placeholders, gchar placeholder)
+{
+	for (size_t i = 0; i < n_placeholders; i++)
+	{
+		if (placeholders[i].placeholder == placeholder)
+			return placeholders[i].replacement;
+	}
+
+	return NULL;
+}
+
+
+/* Replaces %s-style placeholders in a string.
```

Comment:
Just out of curiosity, why not using template-style placeholders and re-using the template replacement code?